### PR TITLE
Add `onNewLine` and `onIdle` run type configurations

### LINKED ISSUE #236: Choose when to lint via extension
Although selene is fast, my computer is slow 😞. These features are useful regardless of performance.

A feature to lint only when:
- the file is saved
- a new line is created
- x amount of time after the user has stopped typing

### PR DESCRIPTION
**Pull request by ok-nick**
This pr closes #236 by adding the `onNewLine` and `onIdle` run type configurations.

I'm not too sure about how I handled detecting new lines, I don't believe there is a better approach?

### CODE REVIEW

**Comment by Kampfkarren** [comment]
If a PR closes an issue, put "Closes #blablabla" so it is done automatically.

File: selene-vscode/src/extension.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import * as selene from "./selene"
 import * as util from "./util"
 import * as vscode from "vscode"
+import * as timers from "timers"
```
Comment (Kampfkarren) [inline selene-vscode/src/extension.ts:4]: ABC order.

File: selene-vscode/src/extension.ts [hunk 2]
```diff
@@ -165,19 +168,34 @@ export async function activate(context: vscode.ExtensionContext) {
         diagnosticsCollection.set(document.uri, diagnostics)
     }
 
+    let lastTimeout: NodeJS.Timeout
     function listenToChange() {
+        const idleDelay = vscode.workspace.getConfiguration("selene").get<number>("idleDelay") as number
```
Comment (Kampfkarren) [inline selene-vscode/src/extension.ts:173]: We only need this on the OnIdle branch, move it there.

File: selene-vscode/src/extension.ts [hunk 3]
```diff
@@ -165,19 +168,34 @@ export async function activate(context: vscode.ExtensionContext) {
...
             case RunType.OnSave:
                 return vscode.workspace.onDidSaveTextDocument(lint)
             case RunType.OnType:
                 return vscode.workspace.onDidChangeTextDocument(event => lint(event.document))
-
+            case RunType.OnNewLine:
+				return vscode.workspace.onDidChangeTextDocument(event => {
+					// Contrary to removing lines, adding new lines will leave the range at the same value hence the string comparisons
+					if (event.contentChanges.some(content =>
+						!content.range.isSingleLine || content.text == "\n" || content.text == "\r\n"
```
Comment (Kampfkarren) [inline selene-vscode/src/extension.ts:183]: `===`

File: selene-vscode/src/extension.ts [hunk 4]
```diff
@@ -165,19 +168,34 @@ export async function activate(context: vscode.ExtensionContext) {
...
+        const idleDelay = vscode.workspace.getConfiguration("selene").get<number>("idleDelay") as number
         switch (vscode.workspace.getConfiguration("selene").get<RunType>("run")) {
             case RunType.OnSave:
                 return vscode.workspace.onDidSaveTextDocument(lint)
             case RunType.OnType:
                 return vscode.workspace.onDidChangeTextDocument(event => lint(event.document))
-
+            case RunType.OnNewLine:
+				return vscode.workspace.onDidChangeTextDocument(event => {
+					// Contrary to removing lines, adding new lines will leave the range at the same value hence the string comparisons
```
Comment (Kampfkarren) [inline selene-vscode/src/extension.ts:181]: I don't understand this, what if we make a new line in the middle of the script?

**Review by Kampfkarren — CHANGES_REQUESTED** [review]

File: selene-vscode/src/extension.ts (hunk 4 above)
Comment (ok-nick) [reply selene-vscode/src/extension.ts:181]: `range.start.line` and `range.end.line` will stay the exact same as a new line is added. While removing lines they are different. `isSingleLine` is a `boolean` to determine if the range is on different lines which still respects what I’ve said above.
If you make a new line in the middle of the script, then the text would either be `\n` or `\r\n`

File: selene-vscode/src/extension.ts [hunk 5]
```diff
@@ -29,6 +30,8 @@ enum Severity {
 enum RunType {
     OnSave = "onSave",
     OnType = "onType",
+    OnNewLine = "onNewLine",
+    OnIdle = "onIdle"
```
Comment (Kampfkarren) [inline selene-vscode/src/extension.ts:34]:
```suggestion
    OnIdle = "onIdle",
```

**Pull request merged**